rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,962 @@
+    """Creates a `Replicator` with the given distribution strategy."""
+
+  def initialize(self) -> List[tf.Operation]:
+    """Initialization to be performed before running any computations.
+
+    In eager mode, it performs any necessary initialization.
+    In graph mode, it creates initialization ops and returns them.
+
+    In graph mode, the returned list will include the initialization ops for any
+    `tf.data.Dataset` iterators created by preceding calls to `prepare_input`.
+
+    Returns:
+      In eager mode, returns an empty list.
+      In graph mode, returns the list of ops to execute.
+    """
+
+  def finalize(self) -> List[tf.Operation]:
+    """Finalization to be performed after the end of all computations.
+
+    In eager mode, it performs any necessary finalization.
+    In graph mode, it creates finalization ops and returns them.
+
+    Returns:
+      In eager mode, returns an empty list.
+      In graph mode, returns the list of ops to execute.
+    """
+
+  def scope(self) -> ContextManager:
+    """Returns a context for model / `Variable` creation.
+
+    Models used within `run` must be created in this scope, so that `Variable`s
+    can be correctly replicated and / or placed on the correct devices.
+
+    N.B. This context is automatically applied within `run`.
+    """
+
+  def prepare_input(
+      self,
+      fn: InputFn[T],
+      replication_mode: InputReplicationMode = InputReplicationMode.PER_WORKER,
+      prefetch_on_device: Optional[bool] = None,
+      enforce_ordering: bool = False) -> InputIterator[T]:
+    """Prepares the input data.
+
+    Depending on the input replication mode, the input function may be called one time
+    or several. We call the result of each call an “input pipeline”. Disjoint subsets of the compute
+    replicas will draw their input data from each input pipeline. See `InputReplicationMode` for
+    more details.
+
+    If `fn` returns an instance of `tf.data.Dataset`, a corresponding `tf.data.Iterator` will be
+    created. Each replica taking its input from this pipeline will execute `Iterator.get_next()`.
+    The `tf.data.Iterator`(s) can be re-initialized by calling the `reinitialize` on the returned
+    `InputIterator` object.
+
+    If `fn` returns a Python callable (e.g. `FIFOQueue.dequeue`), that callable will be executed on
+    each replica taking its input from this pipeline. `InputIterator.reinitialize` will have no effect on
+    this pipeline.
+
+    Args:
+      fn: The input function.
+        The function must take an `InputContext` as the only parameter. It must return either
+        an instance of `tf.data.Dataset` or a parameter-less Python callable that returns a
+        `Tensor` nest.
+      replication_mode: The input replication mode. Defaults to per-worker replication.
+      prefetch_on_device: Indicates whether to prefetch input data to the devices.
+        For now we use a buffer size of 1 for this prefetch.
```

Comment:
There are many theoretical dataset optimizations that one in theory would want to run. Why is prefetching an arg here, rather than something the user does in the input_fn with Dataset? Will this have to be expanded if new performance optimization options are added in the future?

Reply:
The prefetching option here is `prefetch_on_device` which is directly related to the distribution aspect (as opposed to the regular `prefetch` which the user can still have in their input pipeline as usual). I believe the tf.data plan is to not expose this through tf.data optimizations but as an iterator (`MultiDeviceIterator`). Since in this case, DS is creating the iterator for the user, we need to know which type to create.
All that being said, I am leaning towards removing this option and just making that the default behavior. 
cc @rohan100jain thoughts?